Open api

paha77 opened the pull request.

paha77 committed:
Controllers, Security (ROLE_API), Config

paha77 committed:
Controllers, Security (ROLE_API), Config

paha77 committed:
IP parameter, subscribe with or without double opt-in, exception bugfix

paha77 committed:
IP parameter, subscribe with or without double opt-in, exception bugfix

paha77 committed:
API-keys backend, install instructions in README.md, code formatting

paha77 committed:
API-keys backend, install instructions in README.md, code formatting

paha77 committed:
DB-update script

paha77 committed:
DB-update script

paha77 committed:
Unique validator for API-key name field

paha77 committed:
Unique validator for API-key name field

kaufmo commented on Bundle/Controller/OpenApiCategoriesController.php:
```diff
@@ -0,0 +1,323 @@
+        $row = $this->model->getRow($s);
+
+        if (!$row) {
+            throw new NotFoundHttpException('Category not found');
```
$this->newsletterCOmponent->trlKwf('Category not found')

kaufmo commented on Bundle/Controller/OpenApiCategoriesController.php:
```diff
@@ -0,0 +1,323 @@
+    {
+        $this->model = $model;
+        $this->validator = $validator;
+        $this->user = $tokenStorage->getToken() && $tokenStorage->getToken()->getUser() != 'anon.' ?
```
den eingeloggten user bitte nicht im constructor hier nur tokenStorage übergeben

das sollte immer zur laufzeit passieren
anon gibts durch die security eh auch nicht mehr ;)

kaufmo commented on Bundle/Controller/OpenApiCategoriesController.php:
```diff
@@ -0,0 +1,323 @@
+                $subscriber->setLogSource(
+                    ($source = $paramFetcher->get('source')) ?
+                        $source :
+                        trlKwf('Subscribe Open API. API Key: {0}', array($this->getUser()->getUsername())
```
trlKwf bitte nur mit newsletterCOmponent verwenden

kaufmo commented on Bundle/Controller/OpenApiCategoriesController.php:
```diff
@@ -0,0 +1,323 @@
+/**
+ * @Route("/api/v1/open", service="kwc_newsletter.controller.open_api_categories")
+ */
+class OpenApiCategoriesController extends Controller
```
Die 2 Controller bitte in einem Ordner bündeln

kaufmo commented on Bundle/Resources/config/services/model.yml:
```diff
@@ -1,40 +1,50 @@
+    factory: ['Kwf_Model_Abstract', getInstance]
+    arguments: ['KwcNewsletter\Bundle\Model\NewsletterRuns']
+
+  kwc_newsletter.model.newsletter_api_keys:
```
nur weil models hinzugefügt werden bitte nicht das komplette file ändern

kaufmo commented on Bundle/Controller/OpenApiCategoriesController.php:
```diff
@@ -0,0 +1,323 @@
+        }
+
+        // new subscribers from request
+        $subscribers = $paramFetcher->get('subscribers');
```
limit setting bitte zum bundle hinzufügen

kaufmo commented on Bundle/Controller/OpenApiCategoriesController.php:
```diff
@@ -0,0 +1,323 @@
+    public function getUser()
+    {
+        if (!$this->user) {
+            throw new AccessDeniedHttpException('User not logged in');
```
das übernimmt eig eh die security für dich

kaufmo commented on Bundle/Controller/OpenApiSubscribersController.php:
```diff
@@ -0,0 +1,306 @@
+/**
+ * @Route("/api/v1/open", service="kwc_newsletter.controller.open_api_subscribers")
+ */
+class OpenApiSubscribersController extends Controller
```
bitte bestehenden SubscribersCOntroller verwenden und erweitern

beim anmelden gehts eh nur um die Option doubleOptIn auschalten

paha77 committed:
delete $this->getUser() and $this->user, extending subScribersController
for OpenApi, limit mass requests with config, yml format,
Controller/OpenApi* => Controller/OpenApi/*, trlKwf

paha77 committed:
delete $this->getUser() and $this->user, extending subScribersController
for OpenApi, limit mass requests with config, yml format,
Controller/OpenApi* => Controller/OpenApi/*, trlKwf

nsams commented on Bundle/Controller/OpenApi/CategoriesController.php:
```diff
@@ -0,0 +1,341 @@
+    {
+        $this->validator->validateAndThrow($paramFetcher);
+
+        $user = $this->tokenStorage->getToken()->getUser();
```
bitte in apiUser oder openApiUser oder so umbenennen

nsams commented on Bundle/Controller/OpenApi/SubscribersController.php:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    /**
+     * @Route("/subscribers/unsubscribe/{id}", requirements={"id"="[1-9]{1}\d*"})
```
/subscribers/{id}/unsubscribe

nsams commented on Bundle/Controller/OpenApi/CategoriesController.php:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    /**
+     * @Route("/categories/{id}/subscribers", requirements={"id"="[1-9]{1}\d*"})
```
POST .../delete-subscribers?

nsams commented on Bundle/Controller/OpenApi/CategoriesController.php:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    /**
+     * @Route("/categories/{id}/subscribers", requirements={"id"="[1-9]{1}\d*"})
```
.../add-subscribers

nsams commented on Bundle/Controller/OpenApi/CategoriesController.php:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    /**
+     * @Route("/categories/{id}/subscribers", requirements={"id"="[1-9]{1}\d*"})
```
return only ids (can be 10000 - test performance for that case)

paha77 replied:
Changed to ids.  With 12000+ subscribers it takes 7-9 seconds to return the ids (on the test vm)

nsams commented on Bundle/Model/Subscribers.php:
```diff
@@ -19,6 +20,17 @@ class Subscribers extends \Kwf_Model_Db
         )
     );
+    protected $_serialization = array(
+        'gender' => 'user',
```
rename group to openApi

nsams commented on Bundle/Controller/OpenApi/CategoriesController.php:
```diff
@@ -0,0 +1,341 @@
+        $subscribersToCategoryModel = $this->model->getDependentModel('ToSubscribers');
+        $subscribersModel = $subscribersToCategoryModel->getReferencedModel('Subscriber');
+        $s = $subscribersModel->select();
+        $s->whereEquals('id', explode(',', $row->subscriber_ids));
```
that won't work for category with 10000 I fear

nsams commented on Bundle/Controller/OpenApi/SubscribersController.php:
```diff
@@ -0,0 +1,246 @@
+/**
+ * @Route("/api/v1/open", service="kwc_newsletter.controller.open_api.subscribers")
+ */
+class SubscribersController extends SubscribersApiController
```
extending another controller is not a good pattern to use, instead business logic could be in a service used by both controllers

fraxachun commented on README.md:
```diff
@@ -45,3 +45,75 @@ Newsletter Component for Koala Framework
+
+#### Subscriber Open API
+
+* Disable CSRF protection for open api in `config.yml`
```
Instead you could inject this setting by yourself like here: https://gitlab.vivid-planet.com/vivid/kwf-oauth2-server/blob/2.0/DependencyInjection/KwfOAuth2ServerExtension.php

paha77 replied:
It's could be done partially.  (CSRF) config changes can be injected but changes of access_control and adding firewalls in security could not be done in a bundle:

https://github.com/symfony/symfony/issues/22308

paha77 committed:
$user => $openApiUser

paha77 committed:
Changing API endpoint urls

paha77 committed:
Renaming serialization groups: user => openApi

kaufmo committed:
newsletter subscribers can be splitted now with individual source

this can be set in config for every domain or subroot

kaufmo committed:
Merge pull request #83 from koala-framework/newsletter-sources

newsletter subscribers can be splitted now with individual source

paha77 committed:
Moving config into bundle's dependency injection

paha77 committed:
Adding newsletterSource parameter to search and subscribe actions

paha77 committed:
Adding newsletterSource parameter to search and create actions, fixing
serialization groups

paha77 committed:
Returning only subscriber ids on category subscribers search

paha77 committed:
Merge branch 'open-api' of https://github.com/koala-framework/kwc-newsletter into open-api

# Conflicts:
#	Bundle/Controller/OpenApi/CategoriesController.php
#	Bundle/Controller/OpenApi/SubscribersController.php
#	Bundle/DependencyInjection/KwcNewsletterExtension.php
#	Bundle/Model/Categories.php
#	Bundle/Model/Subscribers.php
#	Bundle/Model/SubscribersToCategories.php
#	README.md

paha77 committed:
Subscriber saving in Service

nsams commented on Bundle/DependencyInjection/KwcNewsletterExtension.php:
```diff
@@ -23,4 +24,56 @@ public function load(array $configs, ContainerBuilder $container)
+                    'ignore_paths' => array(
+                        '^/api/v1/subscribers',
+                        '^/api/v1/open',
+                        '^/api/v1/carlog/subscribers',
```
remove this line

nsams commented on Bundle/DependencyInjection/KwcNewsletterExtension.php:
```diff
@@ -23,4 +24,56 @@ public function load(array $configs, ContainerBuilder $container)
+            ));
+        }
+
+        if (isset($bundles['FOSRestBundle'])) { // Add serializer config for FOS rest bundle
```
remove FOSRestBundle config (changes are not specific to this bundle)

nsams commented on Bundle/Controller/OpenApi/CategoriesController.php:
```diff
@@ -0,0 +1,347 @@
+     * @Method("GET")
+     * @View(serializerGroups={"openApi"})
+     */
+    public function getCategorySubscribersAction($id, ParamFetcher $paramFetcher)
```
remove this action completely

paha77 committed:
Removing carlog and fos_rest from DependencyInjection and extending readme

paha77 committed:
Removing category subscribers endpoint

paha77 committed:
Service without paramfetcher and request

paha77 committed:
Removing carlog and fos_rest from DependencyInjection and extending readme

paha77 committed:
Removing usage of expression "subscriber_ids"
